File: 1138_alphabet_board_path/solution.py
# LeetCode 1138 - Alphabet Board Path
# https://leetcode.com/problems/alphabet-board-path/

class Solution:
    def alphabetBoardPath(self, target: str) -> str:
        row, col = 0, 0
        ans: list[str] = []

        def move_to(r: int, c: int) -> None:
            nonlocal row, col
            while row < r:
                ans.append("D")
                row += 1
            while row > r:
                ans.append("U")
                row -= 1
            while col < c:
                ans.append("R")
                col += 1
            while col > c:
                ans.append("L")
                col -= 1

        for ch in target:
            r, c = divmod(ord(ch) - ord("a"), 5)
            if c < col and row == r:
                while col > c:
                    ans.append("L")
                    col -= 1
            else:
                move_to(r, c)
            ans.append("!")
        return "".join(ans)

File: 1138_alphabet_board_path/solution.cpp
// LeetCode 1138 - Alphabet Board Path
// https://leetcode.com/problems/alphabet-board-path/

#include <string>

class Solution {
public:
    std::string alphabetBoardPath(std::string target) {
        int row = 0, col = 0;
        std::string ans;
        for (char ch : target) {
            int r = (ch - 'a') / 5, c = (ch - 'a') % 5;
            while (row > r) { ans += 'U'; --row; }
            while (col > c) { ans += 'L'; --col; }
            while (row < r) { ans += 'D'; ++row; }
            while (col < c) { ans += 'R'; ++col; }
            ans += '!';
        }
        return ans;
    }
};
